File: app/lib/utils/validation.dart
class Validation {
  // Validate email
  String? validateEmail(String email) {
    final emailRegex = RegExp(r'^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$');
    if (!emailRegex.hasMatch(email)) {
      return "Invalid email address";
    }
    return null;
  }

  // Validate name (at least 3 characters, at most 30 characters)
  String? validateName(String name) {
    if (name.length < 3) {
      return "Name must be at least 3 characters long";
    }
    if (name.length > 30) {
      return "Name must be at most 30 characters long";
    }
    return null;
  }

  // Validate password (at least 6 characters, 1 lowercase, 1 uppercase, and 1 special character)
  String? validatePassword(String password) {
    if (password.length < 6) {
      return "Password must be at least 6 characters long";
    }
    if (!RegExp(r'(?=.*[a-z])').hasMatch(password)) {
      return "Password must contain at least one lowercase letter";
    }
    if (!RegExp(r'(?=.*[A-Z])').hasMatch(password)) {
      return "Password must contain at least one uppercase letter";
    }
    if (!RegExp(r'(?=.*[!@#$%^&*(),.?":{}|<>])').hasMatch(password)) {
      return "Password must contain at least one special character";
    }
    return null;
  }
}

File: server/utils/validation.js
const { z } = require("zod");

const emailValidation = z.string().email({ message: "Invalid email address" });

const nameValidation = z
  .string()
  .min(3, { message: "Name must be at least 3 characters long" })
  .max(30, { message: "Name must be at most 30 characters long" });

const passwordValidation = z
  .string()
  .min(6, { message: "Password must be at least 6 characters long" })
  .regex(/(?=.*[a-z])/, {
    message: "Password must contain at least one lowercase letter",
  })
  .regex(/(?=.*[A-Z])/, {
    message: "Password must contain at least one uppercase letter",
  })
  .regex(/(?=.*[!@#$%^&*(),.?":{}|<>])/, {
    message: "Password must contain at least one special character",
  });

const signupValidation = z.object({
  name: nameValidation,
  email: emailValidation,
  password: passwordValidation,
});

const loginValidation = z.object({
  email: emailValidation,
  password: passwordValidation,
});


// Zod middleware to validate request body
const validate = (schema) => (req, res, next) => {
  const result = schema.safeParse(req.body);
  
  if (!result.success) {
    return res.status(400).send({
      message: result.error?.errors[0]?.message,
      errors: result.error?.errors,
    });
  }
  next();
};

module.exports = {
  validate,
  loginValidation,
  signupValidation,
};
